### PR TITLE
Feature: better than-gh

### PR DESCRIPTION
**Pull request by ppezaris**
[Changes reviewed on CodeStream](https://pd-api.codestream.us/r/XyCxqcqdMjbWKP-n/zP1Lc2H4TWKxa9x0QFrG0A?src=GitHub) by pezem2 on Jul 31, 2020

### CODE REVIEW

**Comment by ppezaris** [comment]
FYI i'm starting on making those gear menus work

File: shared/agent/src/providers/github.ts [hunk 1]
```diff
@@ -19,13 +19,18 @@ import {
 	FetchThirdPartyCardsResponse,
 	FetchThirdPartyCardWorkflowRequest,
 	FetchThirdPartyCardWorkflowResponse,
+	FetchThirdPartyPullRequestRequest,
+	FetchThirdPartyPullRequestResponse,
 	GitHubBoard,
 	GitHubCreateCardRequest,
 	GitHubCreateCardResponse,
```
Comment (ppezaris) [inline shared/agent/src/providers/github.ts:26]: ugly

File: shared/agent/src/providers/registry.ts [hunk 2]
```diff
@@ -33,14 +33,18 @@ import {
 	FetchThirdPartyChannelsRequest,
 	FetchThirdPartyChannelsRequestType,
```
Comment (ppezaris) [inline shared/agent/src/providers/registry.ts:34]: comment here

**Comment by ppezaris** [comment]
test

**Comment by ppezaris** [comment]
too heated? WTF?!

**Comment by bcanzanella** [comment]
> 
> 
> too heated? WTF?!

haha, was just testing ;)

**Comment by bcanzanella** [comment]
milestones + projects work now

**Comment by ppezaris** [comment]
so do reactions ;)

File: shared/agent/src/protocol/agent.protocol.providers.ts [hunk 3]
```diff
@@ -330,3 +330,230 @@ export interface EnterpriseConfigurationData {
```
Comment (bcanzanella) [inline shared/agent/src/protocol/agent.protocol.providers.ts]: something else

File: shared/agent/src/protocol/agent.protocol.providers.ts [hunk 4]
```diff
@@ -330,3 +330,230 @@ export interface EnterpriseConfigurationData {
 	baseUrl: string;
 	token: string;
 }
+
+export interface FetchThirdPartyPullRequestRequest {
```
Comment (bcanzanella) [inline shared/agent/src/protocol/agent.protocol.providers.ts:334]: cheese iz good

File: shared/agent/src/protocol/agent.protocol.providers.ts [hunk 5]
```diff
@@ -330,3 +330,230 @@ export interface EnterpriseConfigurationData {
...
+	providerId: string;
+	providerTeamId?: string;
+	pullRequestId: string;
+
+	/**
+	 * in the GitHub world, this is `TeamCodeStream` in https://github.com/TeamCodeStream/codestream
```
Comment (bcanzanella) [inline shared/agent/src/protocol/agent.protocol.providers.ts:340]: manual

File: shared/agent/src/protocol/agent.protocol.providers.ts (hunk 4 above)
Comment (bcanzanella) [inline shared/agent/src/protocol/agent.protocol.providers.ts:334]: q

File: shared/agent/src/managers/documentMarkerManager.ts [hunk 6]
```diff
@@ -38,6 +38,7 @@ import {
 	CSUser
 } from "../protocol/api.protocol";
 import { Functions, log, lsp, lspHandler, Strings } from "../system";
+import * as csUri from "../system/uri";
 import { ReviewsManager } from "./reviewsManager";
```
Comment (bcanzanella) [inline shared/agent/src/managers/documentMarkerManager.ts:42]: single on 42

File: shared/agent/src/providers/registry.ts [hunk 7]
```diff
@@ -19,6 +20,8 @@ import {
 	DisconnectThirdPartyProviderRequest,
 	DisconnectThirdPartyProviderRequestType,
 	DisconnectThirdPartyProviderResponse,
+	ExecuteThirdPartyRequest,
```
Comment (bcanzanella) [inline shared/agent/src/providers/registry.ts:23]: 23

Comment (bcanzanella) [inline shared/agent/src/providers/registry.ts:23]: this is line 23

File: shared/agent/src/providers/registry.ts [hunk 8]
```diff
@@ -1,5 +1,6 @@
 "use strict";
 import { CSMe } from "protocol/api.protocol";
+import { Logger } from "../logger";
```
Comment (bcanzanella) [inline shared/agent/src/providers/registry.ts:3]: line 3

File: shared/agent/src/providers/registry.ts [hunk 9]
```diff
@@ -19,6 +20,8 @@ import {
 	DisconnectThirdPartyProviderRequest,
```
Comment (bcanzanella) [inline shared/agent/src/providers/registry.ts:20]: line 20

File: shared/agent/src/providers/registry.ts (hunk 7 above)
Comment (bcanzanella) [inline shared/agent/src/providers/registry.ts:23]: why is this on line 23?>???>?>?

File: shared/agent/src/managers/markersBuilder.ts [hunk 10]
```diff
@@ -322,6 +327,183 @@ class DefaultMarkersBuilder extends MarkersBuilder {
...
+			startLine: locationAtCurrentCommit.lineStart - 1,
+			endLine: locationAtCurrentCommit.lineEnd - 1,
+			retryWithTrimmedEndOnFailure: true
+		});
+		const remoteDefaultBranchRevisionsPromises = git.getRemoteDefaultBranchHeadRevisions(repoPath, [
+			"upstream",
+			"origin"
+		]);
+		const backtrackShas = [
+			...(await blameRevisionsPromises).map(revision => revision.sha),
```
Comment (bcanzanella) [inline shared/agent/src/managers/markersBuilder.ts:430]: adfadfadadfasdfafddsf

File: shared/agent/src/managers/documentMarkerManager.ts [hunk 11]
```diff
@@ -38,6 +38,7 @@ import {
 	CSUser
 } from "../protocol/api.protocol";
```
Comment (bcanzanella) [inline shared/agent/src/managers/documentMarkerManager.ts:39]: aaaaaa

File: shared/agent/src/git/git.ts [hunk 12]
```diff
@@ -138,7 +138,9 @@ async function setOrFindGitPath(gitPath?: string): Promise<GitLocation> {
 	const gitInfo = await findGitPath(gitPath);
 
 	if (gitInfo.isWsl) {
-		Logger.log(`Git found: ${gitInfo.path} git \u2022 ${Strings.getDurationMilliseconds(start)} ms`)
+		Logger.log(
+			`Git found: ${gitInfo.path} git \u2022 ${Strings.getDurationMilliseconds(start)} ms`
```
Comment (bcanzanella) [inline shared/agent/src/git/git.ts:142]: test comment

File: shared/agent/src/git/git.ts [hunk 13]
```diff
@@ -138,7 +138,9 @@ async function setOrFindGitPath(gitPath?: string): Promise<GitLocation> {
 	const gitInfo = await findGitPath(gitPath);
 
 	if (gitInfo.isWsl) {
-		Logger.log(`Git found: ${gitInfo.path} git \u2022 ${Strings.getDurationMilliseconds(start)} ms`)
+		Logger.log(
```
Comment (bcanzanella) [inline shared/agent/src/git/git.ts:141]: aaa

File: shared/agent/src/git/git.ts [hunk 14]
```diff
@@ -138,7 +138,9 @@ async function setOrFindGitPath(gitPath?: string): Promise<GitLocation> {
 	const gitInfo = await findGitPath(gitPath);
```
Comment (bcanzanella) [inline shared/agent/src/git/git.ts:139]: aa

File: shared/agent/src/git/git.ts [hunk 15]
```diff
@@ -138,7 +138,9 @@ async function setOrFindGitPath(gitPath?: string): Promise<GitLocation> {
...
 
 	if (gitInfo.isWsl) {
```
Comment (bcanzanella) [inline shared/agent/src/git/git.ts:140]: asdasda

File: shared/agent/src/managers/scmManager.ts [hunk 16]
```diff
@@ -5,18 +5,25 @@ import { URI } from "vscode-uri";
...
 	GetBranchesResponse,
 	GetCommitScmInfoRequest,
 	GetCommitScmInfoRequestType,
 	GetCommitScmInfoResponse,
+	GetFileContentsAtRevisionRequest,
+	GetFileContentsAtRevisionRequestType,
+	GetFileContentsAtRevisionResponse,
 	GetFileScmInfoRequest,
 	GetFileScmInfoRequestType,
 	GetFileScmInfoResponse,
```
Comment (bcanzanella) [inline shared/agent/src/managers/scmManager.ts:29]: aaaa

File: shared/agent/src/managers/scmManager.ts [hunk 17]
```diff
@@ -5,18 +5,25 @@ import { URI } from "vscode-uri";
...
 import { Logger } from "../logger";
+import { CodeStreamDiffUriData } from "../protocol/agent.protocol";
 import {
 	BlameAuthor,
 	CoAuthors,
 	CreateBranchRequest,
 	CreateBranchRequestType,
 	CreateBranchResponse,
+	FetchAllRemotesRequest,
+	FetchAllRemotesRequestType,
```
Comment (bcanzanella) [inline shared/agent/src/managers/scmManager.ts:16]: cheese

Comment (bcanzanella) [reply shared/agent/src/managers/scmManager.ts:16]: beer

File: shared/agent/src/managers/markersBuilder.ts [hunk 18]
```diff
@@ -322,6 +327,183 @@ class DefaultMarkersBuilder extends MarkersBuilder {
 	}
 }
 
+class CodeStreamDiffMarkersBuilder extends MarkersBuilder {
```
Comment (bcanzanella) [inline shared/agent/src/managers/markersBuilder.ts:330]: cheese

File: shared/agent/src/managers/markersBuilder.ts [hunk 19]
```diff
@@ -322,6 +327,183 @@ class DefaultMarkersBuilder extends MarkersBuilder {
...
+	private readonly codeStreamDiffUri: CodeStreamDiffUriData;
+
```
Comment (bcanzanella) [inline shared/agent/src/managers/markersBuilder.ts:332]: aaaa

Comment (bcanzanella) [reply shared/agent/src/managers/markersBuilder.ts:332]: a

**Comment by bcanzanella** [comment]
aaaa

File: shared/agent/src/api/codestream/codestreamApi.ts [hunk 20]
```diff
@@ -15,13 +15,13 @@ import { Container, SessionContainer } from "../../container";
 import { Logger } from "../../logger";
 import { isDirective, resolve } from "../../managers/operations";
 import {
+	AddBlameMapRequest,
```
Comment (bcanzanella) [inline shared/agent/src/api/codestream/codestreamApi.ts:18]: 18

File: shared/agent/src/protocol/agent.protocol.reviews.ts [hunk 21]
```diff
@@ -309,7 +309,9 @@ export const CheckPullRequestBranchPreconditionsRequestType = new RequestType<
 >("codestream/review/pr/branch/checkPreconditions");
 
 export interface CheckPullRequestPreconditionsRequest {
-	reviewId: string;
+	reviewId?: string;
+	repoId?: string;
+	branch?: string;
```
Comment (bcanzanella) [inline shared/agent/src/protocol/agent.protocol.reviews.ts:314]: aasdfadf

**Comment by bcanzanella** [comment]
asdfadsf adfs adfsa f
shared/agent/src/protocol/agent.protocol.reviews.ts

**Comment by bcanzanella** [comment]
aaaa
shared/agent/src/managers/scmManager.ts (Lines 79-85)

**Comment by bcanzanella** [comment]
asdad
shared/agent/src/protocol/agent.protocol.ts (Lines 227-234)

**Comment by bcanzanella** [comment]
not in any hunk
shared/agent/src/protocol/agent.protocol.ts (Lines 232-236)
```

export const SetServerUrlRequestType = new RequestType<SetServerUrlRequest, void, void, void>(
	"codestream/set-server"
);```

**Comment by bcanzanella** [comment]
not in any hunk
shared/agent/src/providers/provider.ts (Lines 407-420)
```protected async put<RQ extends object, R extends object>(
		url: string,
		body: RQ,
		headers: { [key: string]: string } = {},
		options: { [key: string]: any } = {}
	): Promise<ApiResponse<R>> {
		await this.ensureConnected();
		return this.fetch<R>(
			url,
			{
				method: "PUT",
				body: JSON.stringify(body),
				headers: { ...this.headers, ...headers }
		```

**Comment by bcanzanella** [comment]
not in a hunk
shared/agent/src/protocol/agent.protocol.codemarks.ts (Lines 72-74)
```	codeBlocks: GetRangeScmInfoResponse[];
	crossPostIssueValues?: CrossPostIssueValues;
}```

**Comment by bcanzanella** [comment]
sdasda
shared/agent/src/protocol/agent.protocol.codemarks.ts (Lines 79-86)
```textDocuments?: TextDocumentIdentifier[];
	entryPoint?: string;
	mentionedUserIds?: string[];
	addedUsers?: string[];
	// codemarks can now be replies
	parentPostId?: string;
	isPseudoCodemark?: boolean;
}```

**Comment by bcanzanella** [comment]
asdf

```
import { Arrays, log, lspProvider, Strings } from "../system";

```
								
[shared/agent/src/providers/bitbucket.ts]([object Object]) (Lines 35-36)

**Comment by bcanzanella** [comment]
asdfaf

```
etRemotePaths,
	ProviderCreatePullRequestRequest,
	ProviderCreatePullRequestResponse,
	ProviderPullRequ
```
								
[shared/agent/src/providers/bitbucket.ts]([object Object]) (Lines 39-42)

File: shared/agent/src/managers/scmManager.ts [hunk 22]
```diff
@@ -5,18 +5,25 @@ import { URI } from "vscode-uri";
...
+	FetchAllRemotesResponse,
 	GetBranchesRequest,
 	GetBranchesRequestType,
 	GetBranchesResponse,
```
Comment (bcanzanella) [inline shared/agent/src/managers/scmManager.ts:20]: asd

**Comment by bcanzanella** [comment]
not in hunk

```
GetReposScmRequest,
```
								
[shared/agent/src/managers/scmManager.ts](https://github.com/teamcodestream/codestream/blob/56b6746d8f80fc52a4506514cbd1babca617a83c/shared/agent/src/managers/scmManager.ts#L44) (Line 44)

File: shared/agent/src/providers/bitbucket.ts [hunk 23]
```diff
@@ -324,6 +326,13 @@ export class BitbucketProvider extends ThirdPartyIssueProviderBase<CSBitbucketPr
 		}
 	}
 
+	@log()
+	getPullRequest(
+		request: FetchThirdPartyPullRequestRequest
+	): Promise<FetchThirdPartyPullRequestResponse> {
+		throw new Error("Method not implemented.");
+	}
```
Comment (bcanzanella) [inline shared/agent/src/providers/bitbucket.ts:334]: why is this here?

**Comment by bcanzanella** [comment]
test test test

```
export interface GetLatestCommittersResponse {
	scm?: {
		[email: string]: string;
	};
	error?: string;
}
```
								
[shared/agent/src/protocol/agent.protocol.scm.ts](https://github.com/teamcodestream/codestream/blob/56b6746d8f80fc52a4506514cbd1babca617a83c/shared/agent/src/protocol/agent.protocol.scm.ts#L275-L280) (Lines 275-280)

File: shared/agent/src/api/slack/slackSharingApi.adapters.ts [hunk 24]
```diff
@@ -825,16 +825,20 @@ export function toSlackPostBlocks(
...
+				if (codemark.id) {
+					actionId = toReplyActionId(counter, codemark, slackUserId);
+					actions.elements.push({
+						type: "button",
+						action_id: actionId,
+						style: "primary",
+						text: {
+							type: "plain_text",
+							text: "View Discussion & Reply"
+						}
```
Comment (bcanzanella) [inline shared/agent/src/api/slack/slackSharingApi.adapters.ts:837]: in hunk

**Comment by bcanzanella** [comment]
half & half

```
	}
			let actionId;
			const actions: ActionsBlock = {
				type: "actions",
				block_id: `codeblock-actions:${counter}`,
				elements: []
			};

			const features = SessionContainer.instance().session.api.features;
			if (features && features.slack && features.slack.interactiveComponentsEnabled) {
				if (codemark.id) {
					actionId = toReplyActionId(counter, codemark, slackUserId);
					actions.elements.push({
						type: "button",
						action_id: actionId,
						st
```
								
[shared/agent/src/api/slack/slackSharingApi.adapters.ts](https://github.com/teamcodestream/codestream/blob/56b6746d8f80fc52a4506514cbd1babca617a83c/shared/agent/src/api/slack/slackSharingApi.adapters.ts#L818-L833) (Lines 818-833)

File: shared/agent/src/protocol/agent.protocol.providers.ts [hunk 25]
```diff
@@ -330,3 +330,244 @@ export interface EnterpriseConfigurationData {
...
+	 */
+	owner?: string;
+	/**
```
Comment (bcanzanella) [inline shared/agent/src/protocol/agent.protocol.providers.ts:343]: test for demo

**Comment by bcanzanella** [comment]
not in a diff hunk

```

export interface AddEnterpriseProviderHostRequest {
	provider: string;
	teamId: string;
	host: string;
	data: { [key: string]: any };
}

export interface AddEnterpriseProviderHostResponse {
	providerId: string;
}

export interface Remo
```
								
[shared/agent/src/protocol/agent.protocol.providers.ts](https://github.com/teamcodestream/codestream/blob/7074d07858271b5d3c401dc592ffd10146642f2b/shared/agent/src/protocol/agent.protocol.providers.ts#L310-L322) (Lines 310-322)

File: shared/agent/src/protocol/agent.protocol.providers.ts [hunk 26]
```diff
@@ -330,3 +330,244 @@ export interface EnterpriseConfigurationData {
...
+
+export interface FetchThirdPartyPullRequestRequest {
+	providerId: string;
+	providerTeamId?: string;
+	pullRequestId: string;
+
+	/**
+	 * in the GitHub world, this is `TeamCodeStream` in https://github.com/TeamCodeStream/codestream
+	 */
+	owner?: string;
```
Comment (bcanzanella) [inline shared/agent/src/protocol/agent.protocol.providers.ts:342]: is diff hunk?

File: shared/agent/src/providers/github.ts [hunk 27]
```diff
@@ -19,13 +21,19 @@ import {
 	FetchThirdPartyCardsResponse,
 	FetchThirdPartyCardWorkflowRequest,
 	FetchThirdPartyCardWorkflowResponse,
+	FetchThirdPartyPullRequestFilesResponse,
+	FetchThirdPartyPullRequestRequest,
```
Comment (bcanzanella) [inline shared/agent/src/providers/github.ts:25]: does it blend?

Comment (bcanzanella) [inline shared/agent/src/providers/github.ts:25]: aaa

File: shared/agent/src/managers/documentMarkerManager.ts [hunk 28]
```diff
@@ -221,6 +222,9 @@ export class DocumentMarkerManager {
 	@lspHandler(FetchDocumentMarkersRequestType)
 	async get(request: FetchDocumentMarkersRequest): Promise<FetchDocumentMarkersResponse> {
 		if (request.textDocument.uri.startsWith("codestream-diff://")) {
+			if (csUri.Uris.isCodeStreamDiffUri(request.textDocument.uri)) {
```
Comment (bcanzanella) [inline shared/agent/src/managers/documentMarkerManager.ts:225]: aaaa

Comment (bcanzanella) [inline shared/agent/src/managers/documentMarkerManager.ts:225]: a

File: shared/agent/package.json [hunk 29]
```diff
@@ -68,7 +68,7 @@
         "uuid": "3.3.2",
         "vscode-languageserver": "5.2.1",
         "vscode-uri": "2.0.3",
-        "write-file-atomic": "3.0.0"
+        "write-file-atomic": "3.0.3"
```
Comment (bcanzanella) [inline shared/agent/package.json:71]: single comment

Comment (bcanzanella) [inline shared/agent/package.json:71]: single

**Comment by bcanzanella** [comment]
no hunk?

```
   "@types/node": "10.1.4",
        "@types/node-fetch": "2.5.0",
        "@types/oauth": "0.9.1",
```
								
[shared/agent/package.json](https://github.com/teamcodestream/codestream/blob/931b2d2b8d1bf60b48c2cd466fc4e81ff7020730/shared/agent/package.json#L81-L83) (Lines 81-83)

File: shared/agent/package.json [hunk 30]
```diff
@@ -68,7 +68,7 @@
         "uuid": "3.3.2",
         "vscode-languageserver": "5.2.1",
         "vscode-uri": "2.0.3",
-        "write-file-atomic": "3.0.0"
```
Comment (bcanzanella) [inline shared/agent/package.json:71]: review

**Comment by bcanzanella** [comment]
aaa

```
   "chai": "4.2.0",
```
								
[shared/agent/package.json](https://github.com/teamcodestream/codestream/blob/931b2d2b8d1bf60b48c2cd466fc4e81ff7020730/shared/agent/package.json#L94) (Line 94)

File: jb/src/main/kotlin/protocols/webview/HostProtocol.kt [hunk 31]
```diff
@@ -147,3 +147,22 @@ class UpdateServerUrlRequest(
 class OpenUrlRequest(
     val url: String
 )
+
+class CompareLocalFilesRequest(
+    val repoId: String,
+    val filePath: String,
+    val headSha: String,
+    val headBranch: String,
```
Comment (bcanzanella) [inline jb/src/main/kotlin/protocols/webview/HostProtocol.kt:155]: aaaa

Comment (bcanzanella) [inline jb/src/main/kotlin/protocols/webview/HostProtocol.kt:155]: aaaa

File: jb/src/main/kotlin/review/ReviewService.kt [hunk 32]
```diff
@@ -1,8 +1,10 @@
 package com.codestream.review
 
 import com.codestream.agentService
+import com.codestream.protocols.agent.GetFileContentsAtRevisionParams
 import com.codestream.protocols.agent.GetLocalReviewContentsParams
 import com.codestream.protocols.agent.GetReviewContentsResult
+import com.codestream.protocols.webview.Context
```
Comment (bcanzanella) [inline jb/src/main/kotlin/review/ReviewService.kt:7]: test

File: jb/src/main/kotlin/agent/AgentService.kt [hunk 33]
```diff
@@ -11,6 +11,8 @@ import com.codestream.protocols.agent.DocumentMarkersParams
 import com.codestream.protocols.agent.DocumentMarkersResult
 import com.codestream.protocols.agent.GetAllReviewContentsParams
 import com.codestream.protocols.agent.GetAllReviewContentsResult
+import com.codestream.protocols.agent.GetFileContentsAtRevisionParams
```
Comment (bcanzanella) [inline jb/src/main/kotlin/agent/AgentService.kt:14]: 14

**Comment by bcanzanella** [comment]
sss

```
s.agent.InitializationOptions
import com.codestream.protocols.agent.Post
import com.codestream.protocols.agent.Review
import com.codestream.protocols.agent.SetServerUrlParams
import com.codestream.protocols.agent.SetServerUrlResult
import com.codestream.protocols.agent.Stream
import com.codestream.settings.ApplicationSettingsSe
```
								
[jb/src/main/kotlin/agent/AgentService.kt](https://github.com/teamcodestream/codestream/blob/9876ee647e42388dde2ef07fe4f4400d4130e0d4/jb/src/main/kotlin/agent/AgentService.kt#L24-L30) (Lines 24-30)

**Comment by bcanzanella** [comment]
single

```
private val logger = Logger.getInstance(AgentService::class.java)
    private var initialization = CompletableFuture<Unit>()

    lateinit var initializeResult: InitializeResult
    lateinit var agent: CodeStreamLanguageServer
    lateinit var remoteEndpoint: RemoteEndpoint
```
								
[jb/src/main/kotlin/agent/AgentService.kt](https://github.com/teamcodestream/codestream/blob/9876ee647e42388dde2ef07fe4f4400d4130e0d4/jb/src/main/kotlin/agent/AgentService.kt#L73-L78) (Lines 73-78)

**Comment by bcanzanella** [comment]
reply<div>line</div>
